### PR TITLE
Pev  new flann demo

### PR DESCRIPTION
**Pull request by pemmanuelviel**
<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by pemmanuelviel** [comment]
The only example currently available for FLANN demonstrates how to use the based matcher.
Using other modes requires some work to figure out at what level to access FLANN functionalities (Index_, GenericIndex, miniflann, etc.) and it appears on forums it's not straightforward for many.
This new sample addresses this problem.
It will also ease the work of potential contributors to FLANN by saving them the burden to write a program to test their modifications in the lib.

File: samples/cpp/tutorial_code/features2D/feature_flann_matcher/SURF_FLANN_search_dataset_Demo.cpp [hunk 1]
```diff
@@ -0,0 +1,270 @@
+// SURF_FLANN_search_dataset_Demo.cpp
```
Comment (alalek) [inline samples/cpp/tutorial_code/features2D/feature_flann_matcher/SURF_FLANN_search_dataset_Demo.cpp:1]: > tutorial_code

This code must belong to some tutorial (it should be mentioned by some tutorial).

Comment (pemmanuelviel) [reply samples/cpp/tutorial_code/features2D/feature_flann_matcher/SURF_FLANN_search_dataset_Demo.cpp]: Moved to sample/cpp/

File: samples/cpp/flann_search_dataset.cpp [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    //-- Show detected matches
+    imshow("Good Matches", img_matches );
+
+    while(1)
+    {
+        int k = waitKey();
+        if (k == 27) // Esc key
+            break;
+    }
+    delete index;
```
Comment (alalek) [inline samples/cpp/flann_search_dataset.cpp:262]: > delete

Lets avoid using of manual memory management.

Please use `std::shared_ptr` approach instead (through `cv::Ptr` / `makePtr` emulation for C++98).

**Pull request merged**